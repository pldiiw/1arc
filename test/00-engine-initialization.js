'use strict';

const engine = require('interpreter').initialize_engine();
const log = console.log;
const hex = n => n.toString(16);
const binToDec = s => parseInt(s, 2);

// Data registers
const data = engine.data;
log(`Number of data registers: ${data.length}`);
data.forEach((reg, i) => {
  log(`${hex(i)} register bit size: ${reg.length}`);
});
log(`Data registers values sum: ${data.reduce((a, reg) => a + binToDec(reg), 0)}`);

// I register
const I = engine.I;
log(`I register bit size: ${I.length}`);
log(`I register value: ${binToDec(I)}`);

// Timer register
const timer = engine.timer;
log(`Timer register bit size: ${timer.length}`);
log(`Timer register value: ${binToDec(timer)}`);

// Sound register
const sound = engine.sound;
log(`Sound register bit size: ${sound.length}`);
log(`Sound register value: ${binToDec(sound)}`);

// Memory
const mem = engine.memory;
log(`Memory bit size: ${mem.length}`);
log(`Memory empty: ${binToDec(mem) === 0}`);

// Program counter
const pc = engine.pc;
log(`Program counter bit size: ${pc.length}`);
log(`Program counter value: ${binToDec(pc)}`);

// Stack pointer
const p = engine.pointer;
log(`Stack pointer bit size: ${p.length}`);
log(`Stack pointer value: ${binToDec(p)}`);

// Stack
const stack = engine.stack;
log(`Stack size: ${stack.length}`);
stack.forEach((v, i) => {
  log(`Stack value ${hex(i)}: ${binToDec(v)}`);
});

// Display
const disp = engine.display;
log(`Display bit height: ${disp.length}`);
log('Number of 64-pixel long display rows: ' +
  disp.reduce((a, v) => a + (v.length === 64 ? 1 : 0), 0));

// Keypad
const kp = engine.keypad;
log(`Keypad bit size: ${kp.length}`);
log(`Keypad values sum: ${kp.reduce((a, v) => a + v, 0)}`);
