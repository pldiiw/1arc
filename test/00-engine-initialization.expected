Number of data registers: 16
0 register bit size: 8
1 register bit size: 8
2 register bit size: 8
3 register bit size: 8
4 register bit size: 8
5 register bit size: 8
6 register bit size: 8
7 register bit size: 8
8 register bit size: 8
9 register bit size: 8
A register bit size: 8
B register bit size: 8
C register bit size: 8
D register bit size: 8
E register bit size: 8
F register bit size: 8
Data registers values sum: 0
I register bit size: 12
I register value: 0
Timer register bit size: 8
Timer register value: 0
Sound register bit size: 8
Sound register value: 0
Memory bit size: 32768
Memory empty: true
Program counter bit size: 12
Program counter value: 0
Stack pointer bit size: 4
Stack pointer value: 0
Stack size: 16
Stack value 0: 0
Stack value 1: 0
Stack value 2: 0
Stack value 3: 0
Stack value 4: 0
Stack value 5: 0
Stack value 6: 0
Stack value 7: 0
Stack value 8: 0
Stack value 9: 0
Stack value A: 0
Stack value B: 0
Stack value C: 0
Stack value D: 0
Stack value E: 0
Stack value F: 0
Display bit height: 32
Number of 64-pixel long display rows: 32
Keypad bit size: 16
Keypad values sum: 0
